fpmms/tools.py

```diff
@@ -594,31 +683,83 @@ def etl(rpc: str, filename: Optional[str] = None) -> pd.DataFrame:
+
+        # Loop through mech addresses in mech_to_info
+        events = []
+        for address, (abi, earliest_block) in mech_to_info.items():
+            if next_start_block_base == 0:
+                next_start_block = earliest_block
+            else:
+                next_start_block = next_start_block_base
+        
+            print(f"Searching for {event_name.value} events for mech {address} from block {next_start_block} to {latest_block}.")
+
+            # parallelize the fetching of events
+            with ThreadPoolExecutor(max_workers=10) as executor:
+                futures = []
+                for i in range(next_start_block, latest_block, BLOCKS_CHUNK_SIZE*SNAPSHOT_RATE):
+                    futures.append(executor.submit(get_events, random.choice(w3s), event_name.value, address, abi, i, min(i + BLOCKS_CHUNK_SIZE*SNAPSHOT_RATE, latest_block)))
+
+                for future in tqdm(as_completed(futures), total=len(futures), desc=f"Fetching {event_name.value} Events"):
+                    current_mech_events = future.result()
+                    events.extend(current_mech_events)
+
+        parsed = parse_events(events)
+
+        contents = []
+        with ThreadPoolExecutor(max_workers=10) as executor:
+            futures = []
+            for i in range(0, len(parsed), 1000):
+                futures.append(executor.submit(get_contents, session, parsed[i:i+1000], event_name))
+
+            for future in tqdm(as_completed(futures), total=len(futures), desc=f"Fetching {event_name.value} Contents"):
+                current_mech_contents = future.result()
+                contents.append(current_mech_contents)
+
+        contents = pd.concat(contents, ignore_index=True)
+
+        full_contents = True
+        if event_name == MechEventName.REQUEST:
+            transformed = transformer(contents)
+        elif event_name == MechEventName.DELIVER:
+            transformed = transformer(contents, full_contents=full_contents)
+
+        events_filename = gen_event_filename(event_name)
+
+        if os.path.exists(events_filename):
+            old = pd.read_csv(events_filename)
+
+            # Reset index to avoid index conflicts
+            old.reset_index(drop=True, inplace=True)
+            transformed.reset_index(drop=True, inplace=True)
+
+            # Concatenate DataFrames
+            transformed = pd.concat([old, transformed], ignore_index=True)
+
+            # Drop duplicates if necessary
+            transformed.drop_duplicates(inplace=True)
+
+        event_to_contents[event_name] = transformed.copy()
+
+    # Store progress
+    tools = pd.merge(*event_to_contents.values(), on=REQUEST_ID_FIELD)
+    store_progress(filename, event_to_contents, tools)
 
 
 if __name__ == "__main__":
-    rpc_ = parse_args()
-    etl(rpc_, DEFAULT_FILENAME)
+    # rpc_ = parse_args()
+    # etl(rpc_, DEFAULT_FILENAME)
+    tools = etl(
+        rpc = 'https://lb.nodies.app/v1/406d8dcc043f4cb3959ed7d6673d311a',
```

Comment:
This won't work, RPCs are expected to be given as a list.
However, why was this change introduced since only the first RPC was utilized?

Reply:
Will correct this. The idea here is to use a pool of RPCs. Will correct this.